File: implementations/ios-sdk/swiftui/Components/AnalyticsEventDisplay.swift
import Combine
import ContentfulOptimization
import SwiftUI

struct AnalyticsEventDisplay: View {
    @EnvironmentObject private var client: OptimizationClient
    @ObservedObject private var store = EventStore.shared

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Analytics Events").font(.headline)
            Text("Events: \(store.events.count)")
                .accessibilityLabel("Events: \(store.events.count)")
                .accessibilityIdentifier("events-count")

            if store.events.isEmpty {
                Text("No events tracked yet")
                    .accessibilityLabel("No events tracked yet")
                    .accessibilityIdentifier("no-events-message")
            } else {
                nonComponentEvents
                componentStatsSection
            }
        }
        .padding()
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("analytics-events-container")
        .onAppear {
            store.subscribe(to: client.eventPublisher)
        }
    }

    private var nonComponentEvents: some View {
        ForEach(Array(filteredEvents.enumerated()), id: \.offset) { index, event in
            let testId = event.componentId.map { "event-\(event.type)-\($0)" }
                ?? "event-\(event.type)-\(index)"
            let desc = eventDescription(event)
            Text(desc)
                .accessibilityLabel(desc)
                .accessibilityIdentifier(testId)
        }
    }

    private var componentStatsSection: some View {
        ForEach(Array(store.componentStats.keys.sorted()), id: \.self) { cid in
            if let stats = store.componentStats[cid] {
                VStack(alignment: .leading) {
                    Text("Count: \(stats.count)")
                        .accessibilityLabel("Count: \(stats.count)")
                        .accessibilityIdentifier("event-count-\(cid)")
                    Text("Duration: \(stats.latestViewDurationMs.map { "\($0)" } ?? "N/A")")
                        .accessibilityLabel("Duration: \(stats.latestViewDurationMs.map { "\($0)" } ?? "N/A")")
                        .accessibilityIdentifier("event-duration-\(cid)")
                    Text("ViewId: \(stats.latestViewId ?? "N/A")")
                        .accessibilityLabel("ViewId: \(stats.latestViewId ?? "N/A")")
                        .accessibilityIdentifier("event-view-id-\(cid)")
                }
                .accessibilityElement(children: .contain)
                .accessibilityIdentifier("component-stats-\(cid)")
            }
        }
    }

    private var filteredEvents: [EventStore.AnalyticsEvent] {
        store.events.filter { $0.type != "component" }
    }

    private func eventDescription(_ event: EventStore.AnalyticsEvent) -> String {
        var desc = event.type
        if let cid = event.componentId { desc += " - Component: \(cid)" }
        if let ms = event.viewDurationMs { desc += " - \(ms)ms" }
        return desc
    }
}

File: implementations/android-sdk/compose/src/main/kotlin/com/contentful/optimization/app/components/AnalyticsEventDisplay.kt
package com.contentful.optimization.app.components

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.contentful.optimization.shared.EventStore

@Composable
fun AnalyticsEventDisplay() {
    val events by EventStore.events.collectAsState()
    val componentStats by EventStore.componentStats.collectAsState()

    Column(
        modifier = Modifier
            .padding(16.dp)
            .testTag("analytics-events-container"),
    ) {
        Text("Analytics Events", fontWeight = FontWeight.Bold)

        val eventsCountText = "Events: ${events.size}"
        Text(
            text = eventsCountText,
            modifier = Modifier
                .testTag("events-count")
                .semantics { contentDescription = eventsCountText },
        )

        if (events.isEmpty()) {
            val noEventsText = "No events tracked yet"
            Text(
                text = noEventsText,
                modifier = Modifier
                    .testTag("no-events-message")
                    .semantics { contentDescription = noEventsText },
            )
        } else {
            val nonComponentEvents = events.filter { it.type != "component" }
            nonComponentEvents.forEachIndexed { index, event ->
                val testId = if (event.componentId != null) {
                    "event-${event.type}-${event.componentId}"
                } else {
                    "event-${event.type}-$index"
                }
                val desc = buildString {
                    append(event.type)
                    event.componentId?.let { append(" - Component: $it") }
                    event.viewDurationMs?.let { append(" - ${it}ms") }
                }
                Text(
                    text = desc,
                    modifier = Modifier
                        .testTag(testId)
                        .semantics { contentDescription = desc },
                )
            }

            componentStats.keys.sorted().forEach { cid ->
                val stats = componentStats[cid] ?: return@forEach
                Column(modifier = Modifier.testTag("component-stats-$cid")) {
                    val countText = "Count: ${stats.count}"
                    Text(
                        text = countText,
                        modifier = Modifier
                            .testTag("event-count-$cid")
                            .semantics { contentDescription = countText },
                    )

                    val durationText = "Duration: ${stats.latestViewDurationMs?.toString() ?: "N/A"}"
                    Text(
                        text = durationText,
                        modifier = Modifier
                            .testTag("event-duration-$cid")
                            .semantics { contentDescription = durationText },
                    )

                    val viewIdText = "ViewId: ${stats.latestViewId ?: "N/A"}"
                    Text(
                        text = viewIdText,
                        modifier = Modifier
                            .testTag("event-view-id-$cid")
                            .semantics { contentDescription = viewIdText },
                    )
                }
            }
        }
    }
}
